programs/marinade-finance/src/stake_system/partial_unstake.rs

```diff
@@ -0,0 +1,209 @@
+use crate::{
+    checks::{check_owner_program, check_stake_matches_validator},
+    stake_system::StakeSystemHelpers,
+};
+
+use anchor_lang::prelude::*;
+use anchor_lang::solana_program::{
+    program::{invoke, invoke_signed},
+    stake::program as stake_program,
+    stake::{self, state::StakeState},
+    system_instruction, system_program,
+};
+
+use crate::{checks::check_address, PartialUnstake};
+
+impl<'info> PartialUnstake<'info> {
+    pub fn process(
+        &mut self,
+        stake_index: u32,
+        validator_index: u32,
+        unstake_amount: u64,
+    ) -> ProgramResult {
+        self.state
+            .validator_system
+            .check_validator_manager_authority(self.validator_manager_authority.key)?;
+        self.state
+            .validator_system
+            .check_validator_list(&self.validator_list)?;
+        self.state.stake_system.check_stake_list(&self.stake_list)?;
+        self.state
+            .check_stake_deposit_authority(self.stake_deposit_authority.key)?;
+        check_owner_program(&self.stake_account, &stake::program::ID, "stake_account")?;
+        self.state
+            .check_stake_deposit_authority(self.stake_deposit_authority.key)?;
+        check_address(self.stake_program.key, &stake::program::ID, "stake_program")?;
+
+        let mut stake = self
+            .state
+            .stake_system
+            .get(&self.stake_list.data.as_ref().borrow(), stake_index)?;
+        if self.stake_account.to_account_info().key != &stake.stake_account {
+            msg!(
+                "Stake account {} must match stake_list[{}] = {}. Maybe list layout was changed",
+                self.stake_account.to_account_info().key,
+                stake_index,
+                &stake.stake_account
+            );
+            return Err(ProgramError::InvalidAccountData);
+        }
+        assert!(
+            stake.is_emergency_unstaking == 0,
+            "already emergency unstaking"
+        );
+
+        let mut validator = self
+            .state
+            .validator_system
+            .get(&self.validator_list.data.as_ref().borrow(), validator_index)?;
+
+        // check that the account is delegated to the right validator
+        check_stake_matches_validator(&self.stake_account.inner, &validator.validator_account)?;
+
+        msg!(
+            "partial unstake {} from account {} current stake {}",
+            unstake_amount,
+            stake.stake_account,
+            stake.last_update_delegated_lamports
+        );
+
+        // only allow partial unstake (avoid code duplication with emergency_unstake)
+        assert!(
```

Comment:
Wrong check. For example unstake_amount may be 0.0000001SOL. unstake_amount >= self.state.stake_system.min_stake && stake.last_update_delegated_lamports - unstake_amount >= self.state.stake_system.min_stake is correct